### PR TITLE
Item Rendering Extensions

### PR DESCRIPTION
**Pull request by Leo40Git**
A rewritten version of my old [item overlay customization API](https://github.com/FabricMC/fabric/pull/916).

Also adds the containing `rendering` library.

### TODO
- [x] Customizing parts of the overlay
  - [x] Count label
  - [x] Item bar
    - [x] Provide API for multiple item bars
  - [x] Cooldown overlay
- [x] Callbacks for before/after overlay rendered
- [x] Javadoc
- [x] Testmod
  - [x] Count label
  - [x] Item bars
  - [x] Cooldown overlay
  - [x] Before/after render callbacks

### CODE REVIEW

**Comment by williambl** [comment]
oooh, remind me to review this later 👀

**Comment by oliviathevampire** [comment]
This looks pretty good! but waiting for other people to review it, you need to update to the latest changes, and I suggets adding a test mod.

File: library/item/item_rendering/build.gradle
Comment (oliviathevampire) [inline library/item/item_rendering/build.gradle]: this should be ``org/quiltmc/qsl/item/rendering/api/client/QuiltItemRenderingExtensions`` instead of ``org/quiltmc/qsl/item/rendering/client/api/QuiltItemRenderingExtensions``

**Comment by 0xJoeMama** [comment]
Can we get BuiltinItemRendererRegistry as well with this PR? Or do you think it's out of scope?

**Comment by Leo40Git** [comment]
> Can we get BuiltinItemRendererRegistry as well with this PR? Or do you think it's out of scope?

I'll do that in a separate PR after this one gets merged.

File: library/item/item_rendering/src/main/java/org/quiltmc/qsl/item/rendering/mixin/client/ItemRendererMixin.java
Comment (williambl) [inline library/item/item_rendering/src/main/java/org/quiltmc/qsl/item/rendering/mixin/client/ItemRendererMixin.java]: This overwrite is going to cause a lot of incompatibilities with fabric mods (
image
) - is there any other way we can do it?

**Review by williambl — CHANGES_REQUESTED** [review]
I am super happy with the API itself. Tests are great too. I only see one problem, the overwrite, since some fabric mods will break with it.

File: library/item/item_rendering/src/main/java/org/quiltmc/qsl/item/rendering/mixin/client/ItemRendererMixin.java
Comment (Leo40Git) [reply library/item/item_rendering/src/main/java/org/quiltmc/qsl/item/rendering/mixin/client/ItemRendererMixin.java]: I made it an `@Overwrite` simply because I couldn't be bothered to redo my old many-parts mixin at the time. I'll try to split it up into `@Inject`s and `@Redirect`s.

Comment (Leo40Git) [reply library/item/item_rendering/src/main/java/org/quiltmc/qsl/item/rendering/mixin/client/ItemRendererMixin.java]: ...also, that third one is 100% copied from my old PR! It has the same comments as well.

**Comment by OroArmor** [comment]
You might hate this, but i think this would fit better under the `rendering` library as an `item` module. Since this entire module is client side and rendering related, i think it fits better there.

**Comment by 0xJoeMama** [comment]
Agreed ^^^

I only just realised this was under the item library.

**Comment by Leo40Git** [comment]
IT HATH BEEN MOVED

**Review by williambl — APPROVED** [review]
Beautiful. Maybe give it a package-info with a high-level overview javadoc? Otherwise, perfect.

**Review by oliviathevampire — APPROVED** [review]

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java
Comment (0xJoeMama) [inline library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java]: To be frank, I think this interface should be split into 2 others. One that holds the static methods, which are general utility methods(you could debate this could be moved to another module as well), and another one with the non-static ones.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/CooldownOverlayRenderer.java
Comment (0xJoeMama) [inline library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/CooldownOverlayRenderer.java:29]: I suppose you could annotate the whole interface with Environment, instead of doing so on its members.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java
Comment (0xJoeMama) [inline library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java]: Do we really need to use ThreadLocal here? I don't think this is accessed by any other thread, other than the Render Thread.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java
Comment (0xJoeMama) [inline library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java]: Does adding `@Unique` make sure that the method never collides? 
Does it salt the method or something? If it doesn't I suppose you could prefix it with quilt$ to ensure nothing weird happens.

Comment (0xJoeMama) [inline library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java]: I also think this should be in a separate class. Might be useful for people.

File: library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/cooldown/FlashingCooldownOverlayRenderer.java
Comment (0xJoeMama) [inline library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/cooldown/FlashingCooldownOverlayRenderer.java:33]: You might want to add an Environment annotation here.

File: library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/cooldown/HiddenCooldownOverlayRenderer.java
Comment (0xJoeMama) [inline library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/cooldown/HiddenCooldownOverlayRenderer.java:31]: Same here.

File: library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/ItemDecorations.java
Comment (0xJoeMama) [inline library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/ItemDecorations.java:34]: Hm. Are these textures stitched into the item texture Atlas? If not, this may cause some performance issues if a lot of mods follow this trend.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java
Comment (0xJoeMama) [inline library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java]: `@Environment`

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/VanillaCountLabelRenderer.java
Comment (0xJoeMama) [inline library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/VanillaCountLabelRenderer.java:43]: I suppose this TODO means you haven't found a way to avoid Immediate right? That's a bit concerning.

Comment (0xJoeMama) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/VanillaCountLabelRenderer.java:43]: I guess vanilla uses Immediate here as well right? I'll have to go look, but I think that's the case.

**Review by 0xJoeMama — COMMENTED** [review]
General Note: Mixin handler methods, are salted at runtime, therefore adding quilt$ in the beginning doesn't really offer a lot in terms of compatibility, while also making the code a bit more annoying to read. Correct me if I am wrong.

I'd also like to note that this review doesn't offer much, in terms of getting the PR approved. It's just some comments I could scratch up.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/CooldownOverlayRenderer.java
Comment (Leo40Git) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/CooldownOverlayRenderer.java:29]: This was purposefully done to prevent any possible "class does not exist" shenanigans on dedicated servers - the interface exists on all sides, but is empty outside of clients.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java
Comment (Leo40Git) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java]: `@Unique` with private methods 100% makes the method never collide, AFAIK.

File: library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/ItemDecorations.java
Comment (Leo40Git) [reply library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/ItemDecorations.java:34]: They are not, I have no idea how to do that

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/VanillaCountLabelRenderer.java
Comment (Leo40Git) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/VanillaCountLabelRenderer.java:43]: Yeah, this is copied from vanilla.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/CooldownOverlayRenderer.java
Comment (0xJoeMama) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/CooldownOverlayRenderer.java:29]: ~~Wouldn't this be the perfect use case for `@EnvironmentalInterface`?~~
Forget this, seems I misunderstood was EnvironmentInterface does.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java
Comment (Leo40Git) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java]: Honestly I should just make this a proper utility class

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java
Comment (0xJoeMama) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java]: image
Yeap javadocs confirms that to be the case.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java
Comment (OroArmor) [inline library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java]: I can hear jellysquid screaming a thousand screams as someone loops this method and submits 10k draw calls, bringing the FPS down to 0. I would remove the draw call and only set the shader if the current shader is different and draw the buffer as it is. Otherwise this is *extremely* inefficient code and bad for rendering.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/QuiltItemRenderingExtensions.java
Comment (OroArmor) [inline library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/QuiltItemRenderingExtensions.java:73]: This `@Nullable` annotation is in a weird place. i would do `@Nullable ItemBarRenderer[]`

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java
Comment (OroArmor) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java]: Its good practice. Other mods might also render twice like Replay mod (Not sure)

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java
Comment (OroArmor) [inline library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/mixin/client/ItemRendererMixin.java]: The line wrapping and formatting on these are a bit weird. I think making your line length longer would fix it. I can't remember if QSL has a limit

File: library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/ItemDecorations.java
Comment (OroArmor) [reply library/rendering/item_rendering/src/testmod/java/org/quiltmc/qsl/rendering/item/test/client/ItemDecorations.java:34]: Texture atlases are extremely complex, and the GUI texture usually used is not atlased. This would be quite an invasive change to Minecraft if this were to be fixed. Not sure the best way to fix it though.

**Review by OroArmor — CHANGES_REQUESTED** [review]
Small nitpicks. I wonder if an accesswidener for the model predicate provider registry could be made in this PR as well.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java
Comment (Leo40Git) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java]: Yeah, I _really_ need to figure out how to make this not horrible (AKA sharing a single `BufferBuilder` and drawing it at the end) while also not making it break should someone set a new shader and not set it back.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/QuiltItemRenderingExtensions.java
Comment (Leo40Git) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/QuiltItemRenderingExtensions.java:73]: `@Nullable ItemBarRenderer[]` -> array itself has no annotation, elements are `@Nullable`
`ItemBarRenderer @Nullable []` -> array itself is `@Nullable`, elements have no annotations (though they're treated as `@NotNull` in the impl)

Comment (OroArmor) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/QuiltItemRenderingExtensions.java:73]: Ah ok. Looks weird to me but makes sense.

File: library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java
Comment (OroArmor) [reply library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/GuiRendererHelper.java]: I would look at the `DrawableHelper` class. I do think it covers most of these methods.

**Comment by Leo40Git** [comment]
I have committed an initial version of the quad batching API (currently untested since I can't run the testmod for some reason).

**Comment by Leo40Git** [comment]
Closing because _man_ I made this overcomplicated as hell. Will bring back later (possibly after extracting the quad batching stuff into its own PR)?

**Comment by maximumpower55** [comment]
Feels like QSL shouldn't be doing this invasive rendering optimizations, quad batching would fit better in its own rendering optimization mod Imo

**Comment by Leo40Git** [comment]
This optimization only applies to items customized via this API - Vanilla items (and items from mods that do _not_ use this API) are unaffected.

**Comment by maximumpower55** [comment]
Ah👍

**Pull request closed without merging**
